### PR TITLE
Implement basic indexing (w/o indexed update or multiple tensor indices)

### PR DESCRIPTION
**Pull request by StrongerXi**
### Summary

Tldr as title.

A few key points, otherwise the documentation in the code offers more explanations
- Indexing essentially creates a special shallowCopy to the indexed tensor that looks only at the indexed region.
- since we handle assignment with SSA, we must re-materialize the indexed node whenever the underlying data changes.

### Test Plan (required)
New unit tests are added
```
make JitNodeTest
make JitEvaluatorTest
make JitTensorTest
```

### CODE REVIEW

**Comment by jacobkahn** [comment]
Methinks this needs rebasing.

**Comment by StrongerXi** [comment]
Rebase.

File: flashlight/fl/tensor/backend/jit/JitTensorBase.h [hunk 1]
```diff
@@ -81,6 +79,10 @@ class JitTensorBase : public TensorAdapterBase {
 
   /**
    * Return the node this JIT tensor represents.
+   * NOTE
+   * 1. `const` w.r.t. the underlying Tensor this represents.
+   * 2. return a node that represents the underlying Tensor, i.e., if there is
+   *    indexing, take care of it.
```
Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/JitTensorBase.h:85]: > take care of it

Can you elaborate a little -- does that mean realize indexing ops into the graph?

Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/JitTensorBase.h:85]: > take care of it

Can you elaborate a little -- does that mean realize indexing ops into the graph?

Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/JitTensorBase.h:85]: > take care of it

Can you elaborate a little -- does that mean realize indexing ops into the graph?

File: flashlight/fl/tensor/backend/jit/JitTensorBase.cpp [hunk 2]
```diff
@@ -7,41 +7,134 @@
...
+  std::shared_ptr<DataStorage> dataStorage_;
+  // TODO consider
+  // 1. using an immutable linked-list here to speed things up
+  // 2. making `std::vector<Index>` into an immutable class and use it as
+  //    shared_ptr (to avoid copying when passing to ViewNodoe), since it's all
+  //    readonly -- JitTensor gives us free immutability for Tensor index.
+  // 3. index merging, e.g., (1, 3)(1, 2) --> (2, 3). Maybe as an optimization
+  //    pass, need to think more.
+  std::vector<std::vector<Index>> indexings_{};
+  Node* oldDataNode_{nullptr}; // nullptr iff no indexings
```
Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/JitTensorBase.cpp:61]: Why use a null pointer for one an an `optional` for the below? Is there a difference between `std::nullopt` and a valid `optional` that has the value `nullptr`? If so, seems a tad error-prone

File: flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        }
+        const auto& tensorIdx = idx.get<Tensor>();
+        // use tensor index's shape as base shape, and then append currently
+        // indexed-result shape onto the base shape (ignore the first dimension
+        // since it's reduced by tensor index)
+        //
+        // In Flashlight, the semantics is a bit different from numpy semantics:
+        // 1. if tensor index has different shape than indexed tensor, the
+        //    tensor index's is flattened first
+        // 2. else the result shape is just the tensor shape -- think `x(x < 0)`
```
Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp:46]: Can we also add a loud TODO here to adjust this behavior?

Would also be interested to see what PT does.

File: flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+namespace fl {
+
+namespace {
+
+Dim ceilDiv(Dim numerator, Dim denominator) {
+  return (numerator + denominator - 1) / denominator;
+}
+
+// TODO refactor with common index logic in OneDnnTensor; maybe put things like
```
Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp:22]: +1. The ArrayFireBackend also has an implementation of this function which is probably less correct, and this is useful for testing too

File: flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp (hunk 3 above)
Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp:46]: Can we also add a loud TODO here to adjust this behavior?

Would also be interested to see what PT does.

File: flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp (hunk 4 above)
Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp:22]: +1. The ArrayFireBackend also has an implementation of this function which is probably less correct, and this is useful for testing too

File: flashlight/fl/tensor/backend/jit/opt/backends/onednn/OneDnnOpFusion.cpp [hunk 5]
```diff
@@ -4,6 +4,7 @@
  * This source code is licensed under the MIT-style license found in the
  * LICENSE file in the root directory of this source tree.
  */
+#include "flashlight/fl/tensor/backend/jit/opt/backends/onednn/OneDnnOpFusion.h"
```
Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/opt/backends/onednn/OneDnnOpFusion.cpp:7]: Duplicate include?

File: flashlight/fl/test/tensor/jit/JitTensorTest.cpp [hunk 6]
```diff
@@ -255,6 +255,101 @@ TEST_F(JitTensorTest, assignmentWithShallowCopyAndCopy) {
...
+  const auto c1 = toJitTensorBase(t1).node();
+  const auto i0 = toJitTensorBase(t0Indexed).node();
+  const auto i1 = toJitTensorBase(t0CopyIndexed).node();
+  const auto i2 = toJitTensorBase(t0ShallowCopyIndexed).node();
+
+  // shallow copy and copy have the right nodes
+  ASSERT_EQ(toJitTensorBase(t0Copy).node(), c0);
+  ASSERT_EQ(toJitTensorBase(t0ShallowCopy).node(), add);
+
+  // graph is properly constructed
```
Comment (jacobkahn) [inline flashlight/fl/test/tensor/jit/JitTensorTest.cpp:311]: 🎉 these tests are awesome

Comment (jacobkahn) [inline flashlight/fl/test/tensor/jit/JitTensorTest.cpp:311]: 🎉 these tests are awesome

**Review by jacobkahn — CHANGES_REQUESTED** [review]
Looks super nice! Some quick questions/changes.

File: flashlight/fl/tensor/backend/jit/JitTensorBase.h (hunk 1 above)
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensorBase.h:85]: Duplicated comment:)

Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensorBase.h:85]: Duplicated comment:)

File: flashlight/fl/tensor/backend/jit/JitTensorBase.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensorBase.cpp:61]: Good point -- no difference. Updated to use `std:nullopt` as well.

File: flashlight/fl/tensor/backend/jit/JitTensorBase.cpp [hunk 7]
```diff
@@ -7,41 +7,134 @@
...
+  //   indexings (recall SSA semantics).
+  // 3. Why `oldDataNode_`?
+  //   When underlying data node changes, not all SharedData objects are
+  //   informed directly, so we lazily perform (2), and use `oldDataNode_` to
+  //   guide us.
+  std::shared_ptr<DataStorage> dataStorage_;
+  // TODO consider
+  // 1. using an immutable linked-list here to speed things up
+  // 2. making `std::vector<Index>` into an immutable class and use it as
+  //    shared_ptr (to avoid copying when passing to ViewNodoe), since it's all
```
[the remark this answers is not in the document]
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensorBase.cpp:56]: Cool, documented!

File: flashlight/fl/tensor/backend/jit/JitTensorBase.h (hunk 1 above)
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensorBase.h:85]: Precisely -- the lazy `viewNode_` update due to underlying data mutation (essentially boils down to a `JitTensorBase::SharedData::updateViewNodeIfNeeded` call).

I actually decided to remove this doc -- client should be agnostic to this lazy materialization detail.

File: flashlight/fl/tensor/backend/jit/JitTensorBase.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensorBase.cpp:61]: Duplicated comment:)

Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensorBase.cpp:61]: Duplicated comment:)

File: flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp (hunk 4 above)
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp:22]: Updated the TODO to mention AF as well.

Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp:22]: Duplicated comment:)

File: flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp (hunk 3 above)
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp:46]: 1. Updated
2. Can confirm PT behaves like Numpy here
3. I removed the second difference -- can't reproduce it anymore, which is a good thing :p. So the only remaining difference is this flattening behavior for > 1D tensor index.

Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp:46]: Duplicated comment:)

**Comment by StrongerXi** [comment]
Rebase & address all feedbacks.

**Review by StrongerXi — COMMENTED** [review]
Some minor fixes on the advanced indexing semantics.

**Pull request closed without merging**